Encryption.md

```diff
@@ -19,216 +19,544 @@
-written to the output stream, followed by the buffer itself.
-
-The column pages (data and dictionary) are encrypted after compression. For each page, 
-the encryption buffer is comprised of an IV, ciphertext and (in case of AES_GCM_V1) of a 
-tag, as described in the Algorithms section. Only the buffer is written to the output 
-stream - not need to write the length of the buffer, since the length (size of the page after
-compression and encryption) is kept in the page headers.
-
-A `crypto_meta_data` field in set in each `ColumnChunk` in the encrypted columns. 
-`ColumnCryptoMetaData` is a union - the actual structure is chosen depending on whether the 
-column is encrypted with the footer key, or with a column-specific key. For the latter, 
-a key metadata can be specified, with a maximal length of 512. Key metadata is a free-form
-byte array that can be used by a reader to retrieve the column encryption key. 
-
-Parquet file footer, and its nested structures, contain sensitive information - ranging 
+Sometimes, a hardware acceleration of AES is unavailable (e.g. in Java 8). Then AES crypto 
+operations are implemented in software, and can be somewhat slow, becoming a performance 
+bottleneck in certain workloads. AES CTR is a regular (not authenticated) cipher. It is 
+faster than GCM cipher, since it doesn’t perform integrity verification and doesn’t 
+calculate the authentication tag. Actually, GCM is a combination of CTR cipher and an 
+authentication layer  called GMAC. For applications running without AES acceleration 
+and willing to compromise on content verification, CTR cipher can provide a boost in 
+Parquet write/read throughput. The second Parquet algorithm encrypts the data content 
+(pages) with CTR, and the metadata (Thrift structures) with GCM. This allows to encrypt/decrypt 
+the bulk of the data faster, while still verifying the metadata integrity and making 
+sure the file has not been replaced with a wrong version. However, tampering with the 
+page data might go unnoticed.
+
+The initialization vectors (IVs) in GCM and CTR ciphers must be unique for each encrypted 
+stream. The CTR IVs are comprised of two parts: a unique nonce (“number used once”) and 
+an initial counter field; both must be passed to the CTR cipher upon stream initialization. 
+The GCM mode creates a CTR cipher internally, and passes an IV comprised of a nonce (passed to 
+the GCM cipher upon stream initialization) and an initial counter field constructed internally 
+by the GCM cipher according to the GCM specification (NIST SP 800-38D).
+Parquet encryption uses the RBG-based (random bit generator) nonce construction as defined in 
+the NIST SP 800-38D document, section 8.2.2. Notice: NIST SP 800-38D uses a term “IV” for what 
+is called “nonce” in the Parquet encryption design.
+
+An input to AES CTR encryption is an encryption key, a 16-byte IV and a plaintext. The output
```

Comment:
I think IVs should be a sub-section.

These two paragraphs introduce what the ciphers expect. The paragraph above should come after these to explain how Parquet encodes the nonce-portion of the IV (or the IV portion of the nonce?) for CTR.